src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+            }
+
+            trackingSpans.Add(new CommentTrackingSpan(selectedSpan, amountToAddToStart, amountToAddToEnd));
+        }
+
+        private static void AddBlockComment(CommentSelectionInfo commentInfo, TextSpan span, List<TextChange> textChanges)
+        {
+            InsertText(textChanges, span.Start, commentInfo.BlockCommentStartString);
+            InsertText(textChanges, span.End, commentInfo.BlockCommentEndString);
+        }
+
+        private static void DeleteBlockComment(BlockCommentSelectionHelper blockCommentSelection, TextSpan spanToRemove,
+            List<TextChange> textChanges, CommentSelectionInfo commentInfo)
+        {
+            DeleteText(textChanges, new TextSpan(spanToRemove.Start, commentInfo.BlockCommentStartString.Length));
+            var endMarkerPosition = spanToRemove.End - commentInfo.BlockCommentEndString.Length;
+            // Sometimes the block comment will be missing a close marker.
+            if (Equals(blockCommentSelection.GetSubstringFromText(endMarkerPosition, commentInfo.BlockCommentEndString.Length),
+                commentInfo.BlockCommentEndString))
+            {
+                DeleteText(textChanges, new TextSpan(endMarkerPosition, commentInfo.BlockCommentEndString.Length));
+            }
+        }
+
+        private class BlockCommentSelectionHelper
+        {
+            private readonly string _text;
+            private readonly ITextSnapshot _snapshot;
+
+            public TextSpan SelectedSpan { get; }
+
+            public IEnumerable<TextSpan> IntersectingBlockComments { get; }
+
+            public IEnumerable<TextSpan> UncommentedSpansInSelection { get; }
+
+            public BlockCommentSelectionHelper(IEnumerable<TextSpan> allBlockComments, SnapshotSpan selectedSnapshotSpan)
+            {
+                _text = selectedSnapshotSpan.GetText().Trim();
+                _snapshot = selectedSnapshotSpan.Snapshot;
+
+                SelectedSpan = TextSpan.FromBounds(selectedSnapshotSpan.Start, selectedSnapshotSpan.End);
+                IntersectingBlockComments = GetIntersectingBlockComments(allBlockComments, SelectedSpan);
+                UncommentedSpansInSelection = GetUncommentedSpansInSelection();
+            }
+
+            /// <summary>
+            /// Determines if the given span is entirely whitespace.
+            /// </summary>
+            public bool IsSpanWhitespace(TextSpan span)
+            {
+                for (var i = span.Start; i < span.End; i++)
+                {
+                    if (!IsLocationWhitespace(i))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+
+                bool IsLocationWhitespace(int location)
+                {
+                    var character = _snapshot.GetPoint(location).GetChar();
+                    return char.IsWhiteSpace(character);
+                }
+            }
```

Comment:
i'm fairly certain we have this helper somewhere.

Reply:
In SyntaxFacts, which isn't in this layer as far as I can tell.  I looked for other helpers but non seemed in an appropriate location / accessible / static

Reply:
we have ITextSnapshotLine.IsEmptyOrWhitespace, would that work?

Reply:
hrmm... you may be more than a line.  i guess there's nothing i can find that helps with a SnapshotSpan